### ISSUE TITLE
Suggest Architecture DI approach

### ISSUE THREAD

**Issue by al1k0s** [scribd/Weaver#82]
Hello, I was trying out Weaver and it worked great. 
But I have a question about injecting the dependencies when I use some architecture, MVC for example. What is the best way to assemble the module while harnessing the Weaver's power? The only obvious solution I found by myself is injecting through the static ```createSomeModule(dependencies:)``` method.
But it won't work with Weaver's annotations and graph validation.
Another problem which appears when architecture is introduced is circular references, giving that Swift doesn't have 2-step initialization like Objc(allocating, initialization) but only 1-step init, circular references usually are resolved through property injection via weak references. I didn't find any mention of this type of injection anywhere.
Can anyone help me?

**Comment by trupin** [comment]
Hello @al1k0s,

Concerning your first question, it is hard to visualize what is your issue without an example. Could you please attach a code sample to demonstrate the problem?

For your second question, weak property injection isn't something that Weaver supports at the moment. You can still do it manually though.  For example:

```swift
final class MovieViewController {
  weak var logger: Logger?
  ...
}

final class HomeViewController {
  // weaver: movieController = MovieViewController
  ...
  func presentMovieController() {
     let controller = dependencies.movieController
     controller.logger = logger
     ...
  }
}
```

Would that solve your problem? If not, could you please provide a code sample demonstrating the problem?

Theo.

**Comment by al1k0s** [comment]
OK, I got my answer. In the following snippet you don't use Weaver with MovieViewController because of not supporting weak injections. 
I was hoping to use Weaver throughout my application but find out that I would still need to make 90% of the injections by myself, because I have lots of modules made with MVC pattern, they reference each other in that particular order M -> C <-> V, that said I will only be able to pass all the dependencies to the Controller via passing them through the view, which is not what I need or via creating static methods which assemble the module but I'll lose most of the cool Weaver functionality.

TLDR: Hope you'll make property injection, that would be amazing.

**Comment by trupin** [comment]
Property injection is already doable (see [documentation](https://github.com/scribd/Weaver#--parameter-annotation)), but they don't support `weak` references at the moment.

For example:
```swift
final class MovieViewController {
  private let dependencies: MovieViewControllerDependencyResolver

  // weaver: delegate <= MovieViewControllerDelegate

  init(injecting dependencies: MovieViewControllerDependencyResolver) {
    self.dependencies = dependencies
  }
}
```

When using `MovieViewController`, its `delegate` will need to be passed down like so:
```swift
dependencies.movieViewController(delegate: self)
```

The current problem when doing so is that `delegate` will be strongly retained, meaning you'll get a retain cycle. 

I could work on adding an annotation to declare `weak` parameters. In which case, the example above could become:
```swift
final class MovieViewController {
  private let dependencies: MovieViewControllerDependencyResolver

  // weaver: delegate <= MovieViewControllerDelegate
  // weaver: delegate.scope = .weak

  init(injecting dependencies: MovieViewControllerDependencyResolver) {
    self.dependencies = dependencies
  }
}
```

I'm ok to file this, if you think that'll solve your problem. I'd put it in the next release.

**Comment by al1k0s** [comment]
Well I mean, thats the only solution we got with swift to deal with cyclic references, and they are common in day-to-day life.

@trupin Would you file this for me, I would much appreciate this being included as I'm much interested in incorporating Weaver into my commercial projects.

**Comment by al1k0s** [comment]
And I would not mind having little guide in the documentation regarding using Weaver with circular references. Right now I'm not sure about next: 
`final class MovieViewController {
  private let dependencies: MovieViewControllerDependencyResolver

  // weaver: view <= MovieView
  // weaver: view.scope = .weak

  init(injecting dependencies: MovieViewControllerDependencyResolver) {
    self.dependencies = dependencies
  }
}`

Does the view property should look like this: `view <= ConcreteAppKitMovieView <- MovieView` to be injected?

The `ideal` situation I look like that:
I have some root app class
It has view dependency
View has the controller dependency
Controller has the view dependency via property injection(aforementioned).
Can the view pass itself as a dependency for a dependency with property injection?

As I see right now you can't do this sort of injection.

so this would be ideal

// ConcreteA conforms to ProtoA
// Arrow means dependency
App <- ConcreteA <- Controller <- weak ProtoA
after initialization ConcreteA with Controller ConcreteA should be injected into the Controller because it conforms to the ProtoA.

Rephrasing:
1) Make possible to classes to register themselves(or make it automatic) for weak injection into the dependencies.
2) Make the weak cyclic injection possible with weak injections and 2-step cyclic injection with the help of (1)

**Comment by al1k0s** [comment]
TLDR: Make weaver great by making possible cyclic injections.
What it takes:
1) Weak property injections
2) Self-registering(to make possible sub-typing with protocols, or make all the protocols which the class conforms to be registered automatically)
3) 2 step graph resolve. 1. Pass the references around. 2. See what weak references you can inject.

**Comment by al1k0s** [comment]
This message contains only my own experience trying adopt Weaver.
You guys are doing a great job, I really like and appreciate the way you imagine DI. 👍
The only bummers I saw was missing cyclic references injection and poor diagnosis messages while I was trying to make Weaver work the way it doesn't work.

**Comment by Logicoeur** [comment]
I second cyclic injections! I ran into issues trying to implement weaver using a VIPER structure with some slight changes. Since we can't use storyboard ViewControllers with weaver, we ended up using our Builder class in the module as a container, and doing all the manual property injection there. Haven't been able to find anything for swift that uses the kind of annotations we love in the Java world, other than Weaver. Great work so far @trupin

**Comment by trupin** [comment]
Hi @al1k0s,

> Would you file this for me, I would much appreciate this being included as I'm much interested in incorporating Weaver into my commercial projects.

No problem, I'll think about the problem in more depth and file it asap.

> Does the view property should look like this: view <= ConcreteAppKitMovieView <- MovieView to be injected?

If `MovieView` is your abstract type of view (protocol or super class), then the property injection would look like this: `view <= MovieView`. Weaver only requires the concrete type to provide a dependency. For an injection, only one type is needed, abstract or concrete.

> Self-registering(to make possible sub-typing with protocols, or make all the protocols which the class conforms to, to be registered automatically)

Would this example illustrate what you're talking about? If so, I agree that this is something we could consider Weaver handling automatically. Can't promise it'll be part of the first step for all this, but I could file a second ticket so that it gets supported at some point.
```
final class Foo {
   // weaver: bar <= BarProtocol
   // weaver: bar.scope = .weak
}
final class Bar {
  // weaver: foo = Foo
  // weaver: foo.scope = .container
}
```
So far this would generate
```
protocol FooDependencyResolver {
   func foo(bar: BarProtocol) -> Foo
}
```
I think what you have in mind is to make Weaver generate
```
protocol FooDependencyResolver {
   var foo: Foo { get }
}
```
And make the injection of `bar` happen internally. Is that right?

**Comment by al1k0s** [comment]
Yes, that what I mean in more concrete example, this is required to make cyclic references possible.

**Comment by trupin** [comment]
Closing this ticket since it resulted in writing #90. Feel free to comment on that ticket for questions/suggestions.

**Comment by bimawa** [comment]
@Logicoeur  I try implement viper with weaver and date with the same solution as custom builders. I'm very excited hear about successful solution with it. I'm currently done infinity loop of dependencies. Can you share with your experience with it? pls ^_^
